File: corejs/core/corepath.js
/*
 * Copyright (C) 2012 Vanderbilt University, All rights reserved.
 * 
 * Author: Miklos Maroti
 */

define([ "core/assert", "core/config" ], function (ASSERT, CONFIG) {
	"use strict";

	// ----------------- PersistentTree -----------------

	var isValidRelid = function (relid) {
		return typeof relid === "string" /* && relid !== "" */;
	};

	var isValidNode = function (node) {
		try {
			var verify = function (text, cond) {
				if( !cond ) {
					throw text;
				}
			};

			verify("object", typeof node === "object");

			verify("relid", (node.relid === null && node.parent === null)
			|| (isValidRelid(node.relid) && typeof node.parent === "object"));

			verify("age", node.parent === null || node.age >= node.parent.age);

			verify("children", node.children === null || typeof node.children === "object");

			verify("lists", node.parent === null || node.children === null
			|| node.parent.children !== null);

			verify("contained", node.parent === null || node.parent.children === null
			|| node.parent.children[node.relid] === undefined
			|| node.parent.children[node.relid].relid === node.relid);
		}
		catch(error) {
			console.log("WRONG NODE: >" + error + "< error", node);
			return false;
		}

		return true;
	};

	var getParent = function (node) {
		ASSERT(isValidNode(node));
		return node.parent;
	};

	var getRelid = function (node) {
		ASSERT(isValidNode(node));
		return node.relid;
	};

	var getLevel = function (node, base) {
		ASSERT(isValidNode(node));
		ASSERT(base === undefined || isValidNode(base));

		var level = 0;
		while( node.parent && node !== base ) {
			++level;
			node = node.parent;
		}

		return level;
	};

	var getRoot = function (node) {
		ASSERT(isValidNode(node));

		while( node.parent ) {
			node = node.parent;
		}

		return node;
	};

	var getAncestor = function (first, second) {
		ASSERT(isValidNode(first) && isValidNode(second));
		ASSERT(getRoot(first) === getRoot(second));

		var a = [];
		do {
			a.push(first);
			first = first.parent;
		} while( first );

		var b = [];
		do {
			b.push(second);
			second = second.parent;
		} while( second );

		// you must be in the same tree
		ASSERT(a[a.length - 1] === b[b.length - 1]);

		var i = a.length - 1;
		var j = b.length - 1;
		while( i !== 0 && a[i - 1] === b[--j] ) {
			--i;
		}

		return a[i];
	};

	var getPath = function (node, base) {
		ASSERT(isValidNode(node));
		ASSERT(base === undefined || isValidNode(base));

		var path = "";
		while( node.parent && node !== base ) {
			if( path === "" ) {
				path = node.relid;
			}
			else {
				path = node.relid + "/" + path;
			}
			node = node.parent;
		}
		return path;
	};

	var joinPaths = function (first, second) {
		ASSERT(typeof first === "string" && typeof second === "string");
		return second ? (first ? first + "/" + second : second) : first;
	};

	var CorePath = function (options) {
		options = CONFIG.copyOptions(CONFIG.corepath, options);

		var maxAge = options.maxAge;
		var agingLimit = options.agingLimit;
		var agingCount = 0;
		var roots = [];

		var detachChildren = function (node) {
			ASSERT(node.children !== null);

			var children = node.children;
			node.children = null;
			node.age = maxAge;

			for( var relid in children ) {
				detachChildren(children[relid]);
			}
		};

		var ageNode = function (node) {
			ASSERT(node.age >= 0 && node.age < maxAge);

			if( ++node.age >= maxAge ) {
				ASSERT(node.parent.children[node.relid] === node);
				delete node.parent.children[node.relid];
				detachChildren(node);
			}
		};

		var ageRoots = function () {
			if( ++agingCount >= agingLimit ) {
				agingCount = 0;

				for( var i = 0; i < roots.length; ++i ) {
					ASSERT(roots[i].age === 0);

					var children = roots[i].children;
					for( var relid in children ) {
						ageNode(children[relid]);
					}
				}
			}
		};

		var actualize = function (node) {
			ASSERT(isValidNode(node));

			if( node.children === null ) {
				node.children = {};

				var parent = actualize(node.parent);
				var child = parent.children[node.relid];

				if( child === undefined ) {
					parent.children[node.relid] = node;
					node.age = 0;
					return node;
				}

				child.age = 0;
				return child;
			}

			while( node.age !== 0 ) {
				ASSERT(node.children !== null);

				node.age = 0;
				node = node.parent;
			}

			return node;
		};

		var createRoot = function () {
			ageRoots();

			var root = {
				parent: null,
				relid: null,
				age: 0,
				children: {}
			};

			roots.push(root);
			return root;
		};

		var getChild = function (node, relid) {
			ASSERT(isValidNode(node) && isValidRelid(relid));

			node = actualize(node);

			var child = node.children[relid];
			if( child === undefined ) {
				ageRoots();

				child = {
					parent: node,
					relid: relid,
					age: 0,
					children: {}
				};

				node.children[relid] = child;
			}

			return child;
		};

		var getDescendant = function (node, head, base) {
			ASSERT(isValidNode(node) && isValidNode(head));
			ASSERT(base === undefined || isValidNode(base));

			var path = [];
			while( head.parent && head !== base ) {
				path.push(head.relid);
				head = head.parent;
			}

			var i = path.length;
			while( --i >= 0 ) {
				node = getChild(node, path[i]);
			}

			return node;
		};

		var getDescendantByPath = function (node, path) {
			ASSERT(isValidNode(node) && typeof path === "string");

			path = path ? path.split("/") : [];

			for( var i = 0; i < path.length; ++i ) {
				node = getChild(node, path[i]);
			}

			return node;
		};

		return {
			isValidRelid: isValidRelid,
			isValidNode: isValidNode,
			getParent: getParent,
			getRelid: getRelid,
			getRoot: getRoot,
			getLevel: getLevel,
			createRoot: createRoot,
			getChild: getChild,
			getAncestor: getAncestor,
			getDescendant: getDescendant,
			actualize: actualize,
			getPath: getPath,
			joinPaths: joinPaths,
			getDescendantByPath: getDescendantByPath
		};
	};

	return CorePath;
});

File: corejava/src/org/isis/webgme/core/CorePath.java
package org.isis.webgme.core;

import java.util.*;

public abstract class CorePath<NODE> {
	public abstract NODE getParent(NODE node);

	public abstract String getRelid(NODE node);

	public abstract NODE createRoot();

	public abstract NODE getChild(NODE node, String relid);

	public abstract boolean isAttached(NODE node);

	public abstract List<NODE> getChildren(NODE node);

	private void printIndent(int indent, StringBuilder builder) {
		while (--indent >= 0)
			builder.append("    ");
	}

	private void printTree(NODE node, int indent, StringBuilder builder) {
		printIndent(indent, builder);
		if (getRelid(node) != null) {
			builder.append("\"");
			builder.append(getRelid(node));
			builder.append("\": ");
		}
		builder.append("{\n");

		if (node instanceof AgedNode) {
			AgedNode anode = (AgedNode) node;

			printIndent(indent, builder);
			builder.append("  age: ");
			builder.append(anode.age);
			builder.append("\n");
		}

		List<NODE> children = getChildren(node);
		if (children.size() > 0) {
			printIndent(indent, builder);
			builder.append("  children: {\n");

			for (int i = 0; i < children.size(); ++i) {
				if (i != 0)
					builder.append(",\n");

				NODE child = children.get(i);
				printTree(child, indent + 1, builder);
			}
			builder.append("\n");
		}

		printIndent(indent, builder);
		builder.append("  }\n");

		printIndent(indent, builder);
		builder.append("}");
	}

	public void printTree(NODE node) {
		StringBuilder builder = new StringBuilder();
		printTree(node, 0, builder);
		System.out.println(builder.toString());
	}

	public int getLevel(NODE node) {
		int level = 0;
		while ((node = getParent(node)) != null)
			level += 1;

		return level;
	}

	public NODE getRoot(NODE node) {
		NODE parent;
		while ((parent = getParent(node)) != null)
			node = parent;

		return node;
	}

	public String getPath(NODE node) {
		ArrayList<String> path = new ArrayList<String>();

		String relid = getRelid(node);
		while (relid != null) {
			path.add(relid);
			node = getParent(node);
			relid = getRelid(node);
		}

		StringBuilder builder = new StringBuilder();

		int i = path.size();
		while (--i >= 0) {
			builder.append('/');
			builder.append(path.get(i));
		}

		return builder.toString();
	}

	public NODE getAncestor(NODE first, NODE second) {
		ArrayList<NODE> a = new ArrayList<NODE>();
		do {
			a.add(first);
			first = getParent(first);
		} while (first != null);
		int i = a.size() - 1;

		ArrayList<NODE> b = new ArrayList<NODE>();
		do {
			b.add(second);
			second = getParent(second);
		} while (second != null);
		int j = b.size() - 1;

		// must have the same root
		assert (a.get(i) == b.get(j));

		while (i != 0 && j != 0 && a.get(i - 1) == b.get(j - 1)) {
			i -= 1;
			j -= 1;
		}

		return a.get(i);
	}

	public boolean isAncestor(NODE node, NODE ancestor) {
		assert (node != null && ancestor != null);

		do {
			if (node == ancestor)
				return true;

			node = getParent(node);
		} while (node != null);

		return false;
	}

	public NODE getDescendant(NODE node, NODE head, NODE base) {
		assert (isAncestor(base, head));

		ArrayList<String> path = new ArrayList<String>();
		while (head != base) {
			path.add(getRelid(head));
			head = getParent(head);
		}

		int i = path.size();
		while (--i >= 0)
			node = getChild(node, path.get(i));

		return node;
	}
}
